File: Arrays/560-Subarray-Sum-Equals-K/Python/560-Subarray-Sum-Equals-K-Optimal.py
"""
Problem: 560. Subarray Sum Equals K
Approach: Optimal - Prefix Sum + Frequency Map

If:
    current_prefix - previous_prefix = k

then the subarray between those two prefix sums has sum k.

Store the frequency of every prefix sum so that all
valid subarrays ending at the current index are counted.

Time Complexity: O(n)
Space Complexity: O(n)
"""

from collections import defaultdict

class Solution:
    def subarraySum(self, nums: list[int], k: int) -> int:
        prefix_sum = 0
        count = 0

        freq = defaultdict(int)
        freq[0] = 1

        for num in nums:
            prefix_sum += num

            count += freq[prefix_sum - k]

            freq[prefix_sum] += 1

        return count

File: Arrays/560-Subarray-Sum-Equals-K/JavaScript/560-Subarray-Sum-Equals-K-Optimal.js
/**
 * Problem: 560. Subarray Sum Equals K
 * Approach: Optimal - Prefix Sum + Frequency Map
 *
 * Time Complexity: O(n)
 * Space Complexity: O(n)
 */

/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var subarraySum = function(nums, k) {
    let prefixSum = 0;
    let count = 0;

    const freq = new Map();
    freq.set(0, 1);

    for (const num of nums) {
        prefixSum += num;

        count += freq.get(prefixSum - k) || 0;

        freq.set(
            prefixSum,
            (freq.get(prefixSum) || 0) + 1
        );
    }

    return count;
};